onboarding_project/squad_pantry_app/admin.py

```diff
@@ -22,11 +31,150 @@ class OrderDishInline(admin.StackedInline):
     formset = BaseOrderDishFormset
     extra = 1
 
+    def formfield_for_foreignkey(self, db_field, request, **kwargs):
+        if db_field.name == "dish":
+            kwargs["queryset"] = Dish.objects.filter(is_available=True)
+        return super(OrderDishInline, self).formfield_for_foreignkey(db_field, request, **kwargs)
+
+    def get_readonly_fields(self, request, obj=None):
+        if obj is None:
+            return self.readonly_fields
+        else:
+            return self.readonly_fields + ('dish', 'quantity')
+
+
+class DishAdmin(admin.ModelAdmin):
+    list_display = ['dish_name', 'dish_type', 'is_available', 'prep_time_in_minutes']
+
+    def has_add_permission(self, request):
+        return request.user.is_kitchen_staff
+
+    def has_change_permission(self, request, obj=None):
+        return request.user.is_kitchen_staff
+
 
 class OrderAdmin(admin.ModelAdmin):
-    inlines = [OrderDishInline,]
+    inlines = [OrderDishInline, ]
+    exclude = ('placed_by', )
+    readonly_fields = ('closed_at', )
+    list_display = ('placed_by', 'status', 'created_at', 'scheduled_time')
+
+    def has_add_permission(self, request):
+        return not request.user.is_kitchen_staff
+
+    def get_urls(self):
+        return [
+            url(
+                r'^(.+)/change/cancel-order/$',
+                self.admin_site.admin_view(self.cancel_order_view),
+                name='cancel_order_view',
+            ),
+        ] + super(OrderAdmin, self).get_urls()
+
+    def cancel_order_view(self, request, object_id, extra_context=None):
+        try:
+            order = Order.objects.get(pk=object_id)
+        except Order.DoesNotExist:
+            self.message_user(request, 'Mentioned Order ID does not exists', messages.ERROR)
+            return HttpResponseRedirect(request.META.get('HTTP_REFERER', '/admin/squad_pantry_app/order'))
+        else:
+            cancelled = order.cancel_order(request.user.id)
+            if cancelled == order.CANCEL_SUCCESS:
+                self.message_user(request, 'Order cancelled', messages.SUCCESS)
+            elif cancelled == order.WRONG_USER:
+                self.message_user(request, 'You do not have permission to cancel this order', messages.ERROR)
+            elif cancelled == order.PROCESSING_ORDER:
+                self.message_user(request, 'Could not cancel, The order is already being processed', messages.ERROR)
+            elif cancelled == order.ORDER_CLOSED_ERROR:
+                self.message_user(request, 'The order is already closed', messages.ERROR)
+
+            return HttpResponseRedirect(request.META.get('HTTP_REFERER', '/'))
+
+    def save_model(self, request, obj, form, change):
+        if not change:
+            obj.placed_by = request.user
+        return super(OrderAdmin, self).save_model(request, obj, form, change)
+
+    def get_readonly_fields(self, request, obj=None):
+        user = request.user
+        if obj is None:
```

Comment:
This looks unnecessarily complicated. Too many nested ifs and elses. Can be refactored as
```py
should_add_status_to_readonly = not user.is_kitchen_staff or (user.is_kitchen_staff and obj.status in obj.CLOSED_ORDERS)

if not obj and should_add_status_to_readonly:
     return <something>
elif should_add_status_to_readonly:
    return <something else>
else:
    return <default>
```

Reply:
Done!